src/Dialect/ONNX/Rewrite.td

```diff
@@ -312,6 +317,22 @@ def RemoveIdentityReshapePattern:  Pat<
   // Check that val has the specified shape.
   [(HasSpecifiedConstantShape $val, $shape)]>;
 
+def RemoveReshapeBeforeMatmul : Pattern<
+ // Reshape is redudant if A and B can be multiplied directly.
+ // E.g. the 1st reshape keep the last dim of A unchanged and the 2nd reshape
+ // produces a type consistent with the ones in A and B.
+ // This pattern sometimes happens to models converted by tf2onnx.
+ (ONNXReshapeOp:$res2 (ONNXMatMulOp (ONNXReshapeOp:$res1 $A, $_), $B), $_),
+ // Just multiply A by B.
+ [(ONNXMatMulOp $A, $B)],
+ [(HasRankOf<3> $A), (HasRankOf<3> $res2),
```

Comment:
It is unfortunate the pattern is specific to tensors with rank equal to 3. I was wondering if there is a way to generalize the pattern, but probably this is OK for now.

Reply:
I concur, it is really a flattening of all but the innermost dim.